addon/components/lazy-mount/component.js

```diff
@@ -0,0 +1,195 @@
+
+  /**
+   * The name of the engine to load and subsequently mount.
+   *
+   * @property name
+   * @type {string}
+   * @public
+   */
+  name: null,
+
+  /**
+   * Optional model that will be passed through to the engine.
+   *
+   * @see https://emberjs.com/api/ember/3.7/classes/Ember.Templates.helpers/methods/mount?anchor=mount
+   *
+   * @property model
+   * @type {any?}
+   * @public
+   */
+  model: null,
+
+  /**
+   * When the engine was loaded successfully, this will then be the name of the
+   * engine. Presence of this field therefore indicates that the engine was
+   * loaded successfully.
+   *
+   * This field is also used by `didReceiveAttrs` for diffing.
+   *
+   * @property loadedName
+   * @type {string?}
+   * @private
+   */
+  loadedName: null,
+
+  /**
+   * If an error occurred while loading the engine, it will be set here.
+   *
+   * @property error
+   * @type {Error?}
+   * @private
+   */
+  error: null,
+
+  /**
+   * While the bundle is being loaded, this property is `true`.
+   *
+   * @property isLoading
+   * @type {boolean}
+   * @private
+   */
+  isLoading: false,
+
+  didReceiveAttrs() {
+    this._super();
+
+    const name = get(this, 'name');
+    assert(`lazy-mount: Argument 'name' is missing.`, name);
+
+    if (name !== get(this, 'loadedName')) {
+      // only load a new engine, if it is different from the last one
+      this.loadEngine(name);
+    }
+  },
+
+  /**
+   * Manages the life cycle of loading an engine bundle and setting the
+   * following properties in accordance:
+   *
+   * - `isLoading`
+   * - `error`
+   * - `loadedName`
+   *
+   * Called by `didReceiveAttrs`.
+   *
+   * @method loadEngine
+   * @param {string} name
+   * @async
+   * @private
+   */
+  async loadEngine(name = get(this, 'name')) {
+    const shouldCancel = this._thread();
+    const engineLoader = get(this, 'engineLoader');
+
+    this.setLoading();
+
+    if (!engineLoader.isLoaded(name)) {
+      try {
+        await engineLoader.load(name);
+        if (shouldCancel()) return;
+      } catch (error) {
+        if (shouldCancel()) return;
+        this.setError(error);
+        return;
+      }
+    }
+
+    this.setLoaded(name);
+  },
+
+  setLoading() {
+    setProperties(this, { loadedName: null, error: null, isLoading: true });
+  },
+  setLoaded(loadedName) {
+    setProperties(this, { loadedName, error: null, isLoading: false });
+  },
+  setError(error) {
+    setProperties(this, { loadedName: null, error, isLoading: false });
+  },
+
+  /**
+   * The following is a really low-fidelity implementation of something that
+   * would be handled by ember-concurrency or ember-lifeline.
+   */
+
+  _threadId: null,
+
+  _thread() {
+    if (Ember.testing) {
+      registerWaiter(this, () => !get(this, 'isLoading'));
+    }
+
+    const threadId = set(this, '_threadId', {});
+    return () =>
+      get(this, 'isDestroyed') ||
+      get(this, 'isDestroying') ||
+      get(this, '_threadId') !== threadId;
```

Comment:
@buschtoens I don't understand this line, because it's always will be equal 

`{} !=== {}`

Reply:
Talked about this on Discord, but for anyone following along at home and also wondering:

Previously `loadEngine` was a restartable ember-concurrency task, meaning that whenever a new instance is started, any old potentially still running instance is cancelled. This is to make sure that, we don't accidentally render stale content, because of async calls, when the input arguments change.

The trick here with `{}` is:

```js
{} === {} // => false

const obj = {};
obj === obj // => true
```

Whenever `loadEngine` is called it starts a new "thread", which then sets `this._threadId` to a new object. So any old still running threads would then compare there own closure `threadId` to the new `this._threadId`, which is then unequal.

Reply:
sound good, thank you bud